# Introducing Machine Breakdowns in DJJSP Environment

## JSSP Environment

In [2]:
from rl4co.envs.scheduling.jssp.env import JSSPEnv
generator_params = {"num_jobs":5 ,
    "num_machines": 6 ,
    "min_processing_time":1 ,
    "max_processing_time": 200
    }
jsspEnv =JSSPEnv(generator_params = generator_params)

In [3]:
jssp_td = jsspEnv.reset(batch_size=[3])
jssp_td 

TensorDict(
    fields={
        action_mask: Tensor(shape=torch.Size([3, 6]), device=cpu, dtype=torch.bool, is_shared=False),
        adjacency: Tensor(shape=torch.Size([3, 30, 30]), device=cpu, dtype=torch.float32, is_shared=False),
        busy_until: Tensor(shape=torch.Size([3, 6]), device=cpu, dtype=torch.float32, is_shared=False),
        done: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.bool, is_shared=False),
        end_op_per_job: Tensor(shape=torch.Size([3, 5]), device=cpu, dtype=torch.int64, is_shared=False),
        finish_times: Tensor(shape=torch.Size([3, 30]), device=cpu, dtype=torch.float32, is_shared=False),
        is_ready: Tensor(shape=torch.Size([3, 30]), device=cpu, dtype=torch.bool, is_shared=False),
        job_done: Tensor(shape=torch.Size([3, 5]), device=cpu, dtype=torch.bool, is_shared=False),
        job_in_process: Tensor(shape=torch.Size([3, 5]), device=cpu, dtype=torch.bool, is_shared=False),
        job_ops_adj: Tensor(shape=torch.Size([3, 

In [4]:
jssp_td ["time"]  # td[time] = [time in batch 0 , time in batch 1 , time in batch 2  ]

tensor([0., 0., 0.])

In [6]:
jssp_td ["busy_until"].shape   #[batch, machine]

torch.Size([3, 6])

## To find till when the machine 3 in batch 1 is busy 

In [8]:

batch_no = 1
machine_no = 3
jssp_td ["busy_until"][batch_no] [machine_no]

tensor(0.)

## DJSSP Environment

- We still do not need change anything

busy_until: specifies until when the machine will be busy-

In [1]:
import torch

from rl4co.envs.scheduling.djssp.env import DJSSPEnv

generator_params = {"num_jobs":5 ,
    "num_machines": 6 ,
    "min_processing_time":1 ,
    "max_processing_time": 200 ,
    "mtbf" : 17 ,
    "mttr" : 4}
env = DJSSPEnv(generator_params=generator_params)

C:\Users\soner\anaconda3\envs\env_rl4co\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
td = env.reset(batch_size=[3])
td

IN BATCH 0 JOB WITH ID 0 not arrived yet
IN BATCH 0 JOB WITH ID 1 not arrived yet
IN BATCH 0 JOB WITH ID 2 not arrived yet
IN BATCH 0 JOB WITH ID 3 not arrived yet
IN BATCH 0 JOB WITH ID 4 not arrived yet
IN BATCH 1 JOB WITH ID 0 not arrived yet
IN BATCH 1 JOB WITH ID 1 not arrived yet
IN BATCH 1 JOB WITH ID 2 not arrived yet
IN BATCH 1 JOB WITH ID 3 not arrived yet
IN BATCH 1 JOB WITH ID 4 not arrived yet
IN BATCH 2 JOB WITH ID 0 not arrived yet
IN BATCH 2 JOB WITH ID 1 not arrived yet
IN BATCH 2 JOB WITH ID 2 not arrived yet
IN BATCH 2 JOB WITH ID 3 not arrived yet
IN BATCH 2 JOB WITH ID 4 not arrived yet


TensorDict(
    fields={
        action_mask: Tensor(shape=torch.Size([3, 6]), device=cpu, dtype=torch.bool, is_shared=False),
        actual_proc_times: Tensor(shape=torch.Size([3, 6, 30]), device=cpu, dtype=torch.float32, is_shared=False),
        adjacency: Tensor(shape=torch.Size([3, 30, 30]), device=cpu, dtype=torch.float32, is_shared=False),
        busy_until: Tensor(shape=torch.Size([3, 6]), device=cpu, dtype=torch.float32, is_shared=False),
        done: Tensor(shape=torch.Size([3, 1]), device=cpu, dtype=torch.bool, is_shared=False),
        end_op_per_job: Tensor(shape=torch.Size([3, 5]), device=cpu, dtype=torch.int64, is_shared=False),
        finish_times: Tensor(shape=torch.Size([3, 30]), device=cpu, dtype=torch.float32, is_shared=False),
        is_ready: Tensor(shape=torch.Size([3, 30]), device=cpu, dtype=torch.bool, is_shared=False),
        job_arrival_times: Tensor(shape=torch.Size([3, 5]), device=cpu, dtype=torch.float32, is_shared=False),
        job_done: Tensor(sh

In [9]:
td["busy_until"][1]

tensor([0., 0., 0., 0., 0., 0.])

In [7]:
check_machine_breakdowns_td = env._check_machine_breakdowns(td)

In [10]:
check_machine_breakdowns_td["busy_until"]

tensor([[0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0.]])

In [3]:
#                       batch    machine_idx
td["machine_breakdowns"][2]       [0]

[{'TIME': 2.182248115539551, 'DURATION': 1.459298014640808},
 {'TIME': 8.896746039390564, 'DURATION': 14.245135307312012},
 {'TIME': 29.159679889678955, 'DURATION': 12.021700859069824},
 {'TIME': 71.04686713218689, 'DURATION': 3.113908290863037}]

In [7]:
td["busy_until"].shape #[batch_no , machine ]

torch.Size([3, 6])

In [8]:
batch_no = 1
machine_idx=2
td["busy_until"][batch_no][machine_idx]

tensor(0.)

## Understanding td["machine_breakdowns"]

In [13]:
batch_no = 1
machine_no = 5
td["machine_breakdowns"][batch_no][machine_no]

[{'TIME': 10.986457824707031, 'DURATION': 9.279200553894043},
 {'TIME': 28.420631408691406, 'DURATION': 6.441053867340088},
 {'TIME': 78.09958791732788, 'DURATION': 0.7334785461425781}]

In [16]:
batch_no = 1
machine_no = 5
breakdown_no = 0
print(td["machine_breakdowns"][batch_no][machine_no][breakdown_no]["TIME"])
print(td["machine_breakdowns"][batch_no][machine_no][breakdown_no]["DURATION"])

10.986457824707031
9.279200553894043


In [ ]:
len(td["machine_breakdowns"][0][1])

In [ ]:
env._check_machine_breakdowns(td)

# TESTING `_check_machine_breakdowns(td)`

In [1]:
import torch

from rl4co.envs.scheduling.djssp.env import DJSSPEnv

generator_params = {"num_jobs":5 ,
    "num_machines": 6 ,
    "min_processing_time":1 ,
    "max_processing_time": 200 ,
    "mtbf" : 17 ,
    "mttr" : 4}
env = DJSSPEnv(generator_params=generator_params)

C:\Users\soner\anaconda3\envs\env_rl4co\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
td = env.reset(batch_size=[3])

IN BATCH 0 JOB WITH ID 0 not arrived yet
IN BATCH 0 JOB WITH ID 1 not arrived yet
IN BATCH 0 JOB WITH ID 2 not arrived yet
IN BATCH 0 JOB WITH ID 3 not arrived yet
IN BATCH 0 JOB WITH ID 4 not arrived yet
IN BATCH 1 JOB WITH ID 0 not arrived yet
IN BATCH 1 JOB WITH ID 1 not arrived yet
IN BATCH 1 JOB WITH ID 2 not arrived yet
IN BATCH 1 JOB WITH ID 3 not arrived yet
IN BATCH 1 JOB WITH ID 4 not arrived yet
IN BATCH 2 JOB WITH ID 0 not arrived yet
IN BATCH 2 JOB WITH ID 1 not arrived yet
IN BATCH 2 JOB WITH ID 2 not arrived yet
IN BATCH 2 JOB WITH ID 3 not arrived yet
IN BATCH 2 JOB WITH ID 4 not arrived yet


## check when the machine breakdown happens and the busy until matrix

In [3]:
td["machine_breakdowns"][0]

{0: [{'TIME': 3.000108480453491, 'DURATION': 10.950005531311035},
  {'TIME': 35.32599139213562, 'DURATION': 1.3882921934127808},
  {'TIME': 62.19732415676117, 'DURATION': 10.096503257751465}],
 1: [{'TIME': 29.66557502746582, 'DURATION': 1.5291308164596558},
  {'TIME': 40.57259452342987, 'DURATION': 1.0042990446090698},
  {'TIME': 80.03481256961823, 'DURATION': 2.711613416671753}],
 2: [{'TIME': 12.460824966430664, 'DURATION': 0.07522933185100555},
  {'TIME': 53.31210486590862, 'DURATION': 1.205381989479065},
  {'TIME': 70.82343074679375, 'DURATION': 0.6587216854095459}],
 3: [{'TIME': 6.723081111907959, 'DURATION': 2.6919376850128174},
  {'TIME': 38.85593867301941, 'DURATION': 0.6645970344543457},
  {'TIME': 57.83890724182129, 'DURATION': 3.2587645053863525}],
 4: [{'TIME': 15.107711791992188, 'DURATION': 8.416229248046875},
  {'TIME': 26.383179187774658, 'DURATION': 3.2661590576171875},
  {'TIME': 89.99576234817505, 'DURATION': 1.8107106685638428}],
 5: [{'TIME': 1.7039726972579956, 

In [4]:
td["busy_until"][0]

tensor([0., 0., 0., 0., 0., 0.])

### In batch 0 machine 0 has the breakdown in time 3.000108480453491 -> in order to test it out, i am manually setting time to 3.000108480453491

In [5]:
after_td = env._check_machine_breakdowns(td)

In [7]:
after_td["busy_until"]

tensor([[13.9501,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000,  0.0000]])

In [ ]:
# it works:)